### ISSUE TITLE
Rethink the Block Input Editor

### ISSUE THREAD

**Issue by cycomachead** [jmoenig/Snap--Build-Your-Own-Blocks#1359]
I'm sure this would be very controversial, but the current UI for editing block parameters in GP is quite nice. To me Snap<em>!</em> feels clunky and a bit overwhelming.

I find the arrows to add inputs more intuitive than the + buttons in Snap<em>!</em>, though it doesn't make it as easy to insert a variable in the middle of a function spec.

Two things I like about GP's editor:
- the dropdown menu to select the "type" of a variable is much more clear than before.
- there's a default value standard (though this may be controversial).

**Comment by jmoenig** [comment]
Thanks! It's always nice to rethink designs in the light of later iterations :-)

**Comment by dekrain** [comment]
My version of Input Dialog:
```javascript
InputSlotDialogMorph.prototype.createSlotTypeButtons = function () {
    // populate my 'slots' area with radio buttons, labels and input fields
    var myself = this, defLabel, defInput,
        oldFlag = Morph.prototype.trackChanges;

    Morph.prototype.trackChanges = false;

    // slot types
    this.addSlotTypeButton('Object', '%obj');
    this.addSlotTypeButton('Text', '%txt');
    this.addSlotTypeButton('List', '%l');
    this.addSlotTypeButton('Number', '%n');
    this.addSlotTypeButton('Any type', '%s');
    this.addSlotTypeButton('Boolean (T/F)', '%b');
    this.addSlotTypeButton('Command\n(inline)', '%cmdRing'); //'%cmd');
    this.addSlotTypeButton('Reporter', '%repRing'); //'%r');
    this.addSlotTypeButton('Predicate', '%predRing'); //'%p');
    this.addSlotTypeButton('Command\n(C-shape)', '%cs');
    this.addSlotTypeButton('Any\n(unevaluated)', '%anyUE');
    this.addSlotTypeButton('Boolean\n(unevaluated)', '%boolUE');

    // arity and upvars
    this.slots.radioButtonSingle = this.addSlotArityButton(
        function () {myself.setSlotArity('single'); },
        "Single input.",
        function () {return myself.fragment.isSingleInput(); }
    );
    this.addSlotArityButton(
        function () {myself.setSlotArity('multiple'); },
        "Multiple inputs (value is list of inputs)",
        function () {return myself.fragment.isMultipleInput(); }
    );
    this.addSlotArityButton(
        function () {myself.setSlotArity('upvar'); },
        "Upvar - make internal variable visible to caller",
        function () {return myself.fragment.isUpvar(); }
    );
    this.addSlotArityButton( /*here*/
        function () {new DialogBoxMorph(myself,function(res){myself.setSlotType('%'+res)},myself.world()).prompt('Select spec','s',myself.world())},
        'Select spec',
        function() {return false}
    )

    // default values
    defLabel = new StringMorph(localize('Default Value:'));
    defLabel.fontSize = this.slots.radioButtonSingle.fontSize;
    defLabel.setColor(new Color(255, 255, 255));
    defLabel.refresh = function () {
        if (myself.isExpanded && contains(
                ['%s', '%n', '%txt', '%anyUE'],
                myself.fragment.type
            )) {
            defLabel.show();
        } else {
            defLabel.hide();
        }
    };
    this.slots.defaultInputLabel = defLabel;
    this.slots.add(defLabel);

    defInput = new InputFieldMorph(this.fragment.defaultValue);
    defInput.contents().fontSize = defLabel.fontSize;
    defInput.contrast = 90;
    defInput.contents().drawNew();
    defInput.setWidth(50);
    defInput.refresh = function () {
        if (defLabel.isVisible) {
            defInput.show();
            if (myself.fragment.type === '%n') {
                defInput.setIsNumeric(true);
            } else {
                defInput.setIsNumeric(false);
            }
        } else {
            defInput.hide();
        }
    };
    this.slots.defaultInputField = defInput;
    this.slots.add(defInput);
    defInput.drawNew();

    Morph.prototype.trackChanges = oldFlag;
};
```

**Comment by jmoenig** [comment]
Interesting, can you post a screenshot of this? Thanks!

**Comment by dekrain** [comment]
![block input dialog screen](https://cloud.githubusercontent.com/assets/19158291/17638536/e8bf7a36-60ea-11e6-880f-321fb3eb4301.png)

**Comment by jmoenig** [comment]
Thanks! What does "select spec" do? Looks like you can specify the internal input-slot symbol here, so it is a kind of shortcut, right?

**Comment by dekrain** [comment]
select spec - write 'clr', spec is '%clr'; write 'mult%s', spec is '%mult%s'

**Comment by brianharvey** [comment]
I guess I wouldn't hate it if the rightmost + were an arrowhead, but I really vote for keeping the present UI.  The plus signs are explained both in the manual and in the Make a Block help screen (one reason I prefer teaching using the button first, rather than the right-click shortcut).  And I wouldn't want to lose ≤ as an example!

**Comment by cycomachead** [comment]
Yeah, I'm not convinced there's an one correct design. I _do_ like the idea of a hybrid "+" with a right arrow at the end. I think that could work well. Since Scratch, I thought a bit more about this - but this is all kind of just ideas.

However, the crux of this issue way more about the dialog box itself. The array of options is, IMO, confusing. Plus certain aspects, like menu inputs are virtually undiscoverable. We are missing some potential input types (colors, primarily) that are more easy to extend as a dropdown than as this large grid.

One thing I was thinking of was if there were a dropdown to select a type, and then one option of that menu would be "Advanced Options" or something which could give users more control - such as default inputs, variadic-ness, or to be unevaluated (if applicable). 

There are two reasons I'd like to separate out the process of defining a type and of setting its options.
* Setting an input type is fairly frequent, but most users don't adjust more than the defaults
* Separating this would allow us to build more complex options without making the normal actions more complex.

I would image that each block input would have a dropdown menu (like GP) and one of this options could be "More Controls" or something. It would pop open a window similar to the current one. However, in my mind we could redesign the current window to have a more directed flow. On the left side we could have a menu / list of the different types of inputs, and upon selecting a type the right hand pane would reveal the options applicable to that input.

**Comment by brianharvey** [comment]
We already make the common case easy -- the first dialog you see just has a slot for the name!  Renenber that this isn't Java; non-procedure types are entirely advisory.  So most people will never see the long dialog.

The arrangement of the types in two dimensions is pedagogic -- see the manual.

Of course I agree about the pulldown option; that shouldn't be hidden in a context menu.  It should be another of the variant choices at the bottom.

And yes, when we have first class colors we'll have to find room for them in the array somehow.  Maybe we'll have other object types to add, too, so you can see the difference between a sprite input and a costume input (a little Alonzo for the latter?).

Also, taking off my Snap<i>!</i> hat and putting on my BJC hat, let me note that any change that's visible to users means that curriculum writers have to find all the pictures in the curriculum and replace them, or that teachers get horribly confused.  We programmers can cope with programs that don't quite match the documentation, but it's harder for regular people.  So I think any proposed change had better be a *lot* better to justify the Flag Day it'd cause.

**Comment by cycomachead** [comment]
> We already make the common case easy -- the first dialog you see just has a slot for the name! Renenber that this isn't Java; non-procedure types are entirely advisory. So most people will never see the long dialog.

I don't think "most people" will never see the dialog. BJC uses input types all over the place, and CS10 / Dan definitely encourage using it all the time. (I feel like i could quote SICP about programs being for humans to read, but I feel like you probably have a good reason why types don't do that.  :) )

The other thing we could do with such a refactoring is even eliminate the need for an extra modal dialog. Why not just click on the text / name of the variable and get an inline edit field? When you click away or hit enter things would be saved.

> The arrangement of the types in two dimensions is pedagogic -- see the manual.

Yeah, i've seen the grouping, and it does make sense, but I'm not convinced it's helpful, because the "reports" namely list and numbers aren't in that column and those are probably the two most common inputs people will choose to use.

> Of course I agree about the pulldown option; that shouldn't be hidden in a context menu. It should be another of the variant choices at the bottom.

So, this one of the problems I think of this dialog. The bottom controls _look_ like they are modifiers of the top section, but they aren't, at least not always. A dropdown is mutually exclusive with any input type, as is an upvar. 


> Also, taking off my Snap! hat and putting on my BJC hat, let me note that any change that's visible to users means that curriculum writers have to find all the pictures in the curriculum and replace them, or that teachers get horribly confused. We programmers can cope with programs that don't quite match the documentation, but it's harder for regular people. So I think any proposed change had better be a lot better to justify the Flag Day it'd cause.

Yeah, I don't think this will happen any time soon -- the current UI is functional, just challenging (I think). Eventually, I hope we can get to procedurally generated blocks/images in the curriculum and not worry about this at all.

**Comment by brianharvey** [comment]
I'll grant you upvars, but a pulldown does modify a type -- compare POINT IN DIRECTION with any of the Sensing ones.

> Why not just click on the text / name of the variable and get an inline edit field?

I guess that would be okay.  It would have to have a clickable arrowhead to open the long dialog.  Although actually I like having text/variable in the short dialog because sometimes I'll enter all of a complicated prototype in the initial make-a-block dialog, e.g., "for i = start to end action" and then change some of the words to variables later.  (I don't like the % thing, just personal taste.)

> Eventually, I hope we can get to procedurally generated blocks/images in the curriculum and not worry about this at all.

Unlikely.  You're not going to get things like procedurally generated animated GIFs.  Or stuff from Snap<i>!</i> annotated in Photoshop.

**Comment by cycomachead** [comment]
> I'll grant you upvars, but a pulldown does modify a type -- compare POINT IN DIRECTION with any of the Sensing ones.

Yeah, I guess I was thinking about that too -- though right now this isn't possible for custom blocks. It would be cool if dropdowns could be of any type (like a command block), but I suppose that will come if we ever get reporters as the defining list.



> You're not going to get things like procedurally generated animated GIFs. Or stuff from Snap! annotated in Photoshop.

Yeah, I guess I really just mean reducing the effort. Once we make it easy to save a screenshot of a script, the GIF isn't much harder to do, though some stuff will always be a bit manual.

**Comment by brianharvey** [comment]
>  though right now this isn't possible for custom blocks.

Is too:
![foo](https://cloud.githubusercontent.com/assets/3885074/18037960/818062d4-6d44-11e6-8f12-fbc37d1d6bee.png)
:-)

**Comment by cycomachead** [comment]
Oh, cool!

I couldn't get that to work on time, but that's good to know!
ᐧ

--
Michael Ball | michaelball.co
